### ISSUE TITLE
sql: validity of sub-queries in special query contexts

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#41078]
tldr: this issue spells out the semantic rules around the `[...]` syntax, CTEs, mutations and hoisting of subqueries. **Comments welcome**

Topics covered:

- equivalence between `[...]` and WITH
- CTE hoisting
- mutations in CTEs and the pg/SQL restriction
- the case of EXPLAIN

### Intro

CockroachDB's dialect allows a few _syntactic_ constructs that are not possible in PostgreSQL. The question then arises what is their _semantics_, if any. 

For example, we like things like `select * from [explain ...] where field=...`, which postgres does not support, and that _seems_ at first to be well defined.  But when what happens with `select * from [explain select from [explain ...]]`? 

(summary from below: easy peasy, just compute the `explain` output depth first post-order)

We would also like things like `table [update ... returning ...] union all table [update ... returning ...]` but it looks suspicious: beyond the apparent clarity of this syntax, is it valid to optimize the _execution_ of this query to run the two updates concurrently?

Then what about `select * from [insert into x table [insert into y values... returning ...] returning...]`?

(summary from below: _it depends_)

### Terminology

- common table expression (_CTE_): the `xxx` part in `WITH a AS (xxx)`
- CTE _alias_: the `xxx` part in `WITH xxx AS (SELECT ...)`
- _scalar subquery_: a valid query enclosed in `(...)` in a scalar position, e.g. `xxx` in `SELECT 1 IN (xxx)`
- _relational subquery_: a valid query enclosed in `(...)` in a relational position, e.g. `xxx` in `SELECT 1 FROM (xxx)`
- _statement source_: a statement enclosed in `[...]` e.g. `SELECT 1 FROM [xxx]`
- _correlated subquery_: a subquery that references columns defined out of it. For example, in `SELECT (SELECT x+y) FROM xy` where `x` and `y` are columns of `xy`, the subquery `SELECT x+y` is correlated.

### Clarification on crdb's statement sources

The "statement source" syntax `[...]` is intended to be equivalent to relational subqueries and CTEs. 

In fact it exists in CockroachDB mainly for historical reasons, as it was introduced at a time when CockroachDB did not support CTEs.

For technical reasons specific to how the SQL parser works, it's not possible to allow all the statement types that may be interesting to use as relational expression, inside `(...)`. However, these statement types _can_ be included syntactically as a CTE, so once the CTE syntax became available there was no unique reason to keep `[...]` other than, perhaps, conciseness and backward-compatibility.

### Mapping statement sources to CTE

In fact, it's _always_ possible to perform a text-level substitution of all uses of `[...]` and replace it with `WITH ...`, for example:

| Before | After | Notes |
|---|---|---|
| `table [xxx]` | `with a as (xxx) table a` | define CTE just above, replace statement source by CTE alias
| `select * from [xxx]` | `with a as (xxx) select * from a` | ditto |
| `insert into t table [xxx]` | `insert into t with a as (xxx) table a` | ditto, note we're not defining the CTE completely at the top yet. we'll discuss top-level CTEs later. |
| `table [select * from [xxx]]` | with a as (xxx), b as (select * from a) table b` | transform statement source to CTE in the deeper level first, then the outer level |

In other words **the semantics of a statement source are those of a CTE defined at the same level**, there is nothing special with them that a CTE cannot do.

### CTEs are actually more powerful

In fact, CTEs do something _more_ which statements sources cannot: it's possible to refer to a CTE alias more than once. For example I can write:

`with a as (select 1) select * from a, a`

and the results of `(select 1)` will be cross-joined with themselves.

To the naive reader, this may look&feel like we're just "pasting" the CTE definition to every place where the CTE alias is listed, so that it looks possible to reverse map a CTE to a statement source. But that's incorrect, because of this example:

`with a as (insert into t values(1) returning *) select * from a, a`

In this case, PostgreSQL's semantics (and the SQL standard, actually) _guarantee_ that the INSERT is only performed once. 

This means that a CTE also conceptually "stores" the intermediate results of its query somewhere and reuses the stored results every time the CTE alias is used. It does not re-compute the query every time. This is not possible to achieve with statement sources only.

### Statement sources are not intended to be more powerful

The `[...]` syntax is a CockroachDB extension and we do not have much evidence it's used in the wild. We are attached to it because it makes our SQL tests much shorter and more convenient to it, but we don't have strong requirements on them other than "don't make them do less than they could do in CockroachDB 2.0".

In particular **the statement source syntax was introduced when CockroachDB did not support correlated subqueries**. This did not work in 2.0, PostgreSQL is not providing it, and nobody is asking for it. Therefore we do not have any requirement that statement sources work when the query inside is correlated. Remember this as we'll use this fact below.

### Do not pay too much attention to statement sources

Because statement source syntax can always map to CTEs, we may as well consider that statement sources are transformed to a CTE *at the same scope level* very early during query transformation, before semantic analysis. 

This enables reasoning about the semantics of SQL without taking statement source syntax into consideration. We'll simply focus on CTEs from now on.

### CTE hoisting

Consider the two following SQL queries:

```sql
select x+1 from (with a as (select 1 as x) table a)
```

and

```sql
with a as (select 1 as x) select x+1 from table a
```

What's the difference?  In the first case the CTE is _nested_, in the latter case the CTE is at the _top level_ of the query.

For various reasons related to performance, it is desirable during query optimization to _hoist_ nested CTEs from "inside" a query to the toplevel, or close to it.

The question then arises, when is this transformation valid?

In the example above, the two queries are equivalent and the transformation is possible. However, in the following case, not so much:

```sql
select 1 = (with a as (select x) table a) from t
```

(where `x` is a column from `t`.

In this example, it's not possible to transform this to:
```sql
with a as (select x)
select 1 = (table a) from t
```

because the scope of `t` is limited to `select...` and therefore the `with ...` becomes invalid.

In general, hoisting a *correlated CTE* is problematic. To address this a query optimizer has two tools:

1. as much as possible, a nested CTE can be "inlined" in the relational query tree, and the need for the WITH... aliasing disappears entirely
2. in _theoretical_ cases where a CTE cannot be hoisted, it may be necessary to define a `LET ... IN ...` relational expression node that preserves the CTE constructs in the final query plan. 

However, SQL databases really do not like case 2 above, and we'd like to be in a position to _always_ hoist nested CTEs, or de-correlate them to inline them in the surrounding context.

### The problem with mutations

*In general, mutations cannot always be decorrelated when they are nested in a larger query context.*  (for various technical and theoretical reasons).

So what do we have here:
1. it's possible to have a SQL mutation inside a CTE (e.g. `with a as (update...)`)
2. we really want to decorrelated CTEs or hoist them to the top level, and not get stuck with nested CTEs in the final plan
3. not all mutations can be decorrelated

*what happens if there's a correlated + nested + mutation CTE in a query?**

This looks like we'd be stuck?

### PostgreSQL to the rescue

Well.... PostgreSQL and the SQL standard rescue us on this point:

**These SQL dialects which we aim to support only support mutations in top-level CTEs**. 

They don't even have to mention "decorrelated" - they use "top level" which is even more powerful. It's not possible to have a correlated query at the top level.

In other words, if a client asks them to run SQL that contains UPDATE etc in a CTE that's nested, they tell them "the SQL syntax is not correct".

This _semantic limitation_ that our upstream techs are forcing on their users is good for us, because we can keep it and it makes our lives easier.

With this limitation in place, we can simply reject a query like this:

`select * from (with a as (update .... returning))` with an error that sounds like "cannot use UPDATE in a CTE that's not at the top level" or something like that.

### Mutations in statement sources

To the astute reader the sections above land into a contradictory, unsound spot:

- at the beginning we say "It's desirable to support `select * from [insert into ... returning ...]`
- then we say "It's always possible to transform statement sources to a CTE at the same level" so that query would be `select * from (with a as (insert ...) table a)`
- then we say "mutations in nested CTEs should be disallowed" which means this desirable query becomes invalid?

That's no good?!

Well to square the circle we can reliably fix this by replacing the rule "no mutations in nested CTEs" to either:

A. "no mutations in CTEs that cannot be hoisted + the remaining mutations are always hoisted", or
B. "no mutations in correlated CTEs + the remaining mutations are always hoisted"

("no mutations" = "produce a planning error if a client tries that")

Either of these two rules will guarantee that a _nested_ mutation in a statement source syntax or CTE definition can become hoisted at the top level during query transformation.

Both of these two rules are more generous to clients than the base rule from PostgreSQL / SQL standard, so we're not losing anything regarding expressiveness by adopting them.

The question then becomes, which of the two variants to choose.

I (@knz) personally recommend going for variant B. 

The problem with variant A is that it opens the conversation of _which_ nested queries containing mutations can be hoisted (or de-correlated and then hoisted), and how. This is a super hard discussion and we'd be trying to solve a problem that nobody else in the pg world is otherwise trying to solve. 

Variant B of the rule is much simpler to implement as this determination can be done very early during query analysis, during scope analysis to be specific (optbuilder in CockroachDB). It's also much simpler to explain and teach, and to for users to get an intuition about.

### Example

Consider:

```sql
select * from [insert into t values(x) returning *] from xy
```

This transforms to
```
select * from (with a as (insert into t values(x) returning *)) from xy
```

And then we see that `insert into t values (x)` is correlated and so by the rule above the query planning fails with "INSERT not allowed in correlated CTE".

Another example:

```sql
select * from [insert into t [insert into u values (1) returning *] returning *]
```

This transforms to
```sql
select * from (
  with a as (insert into u values (1) returning *),
  b as (insert into t table a returning *)
  table b
)
```

Here the mutations are not correlated, and so all is well. CTE hoisting will pull them at the top level safely:

```sql
with a as (insert into u values (1) returning *),
b as (insert into t table a returning *)
select * from b
```

### Multiple mutations in CTEs

What we have so far is that **all nested mutations eventually make their way to a top level CTE definition** via tree transforms.

In fact it's valid pg SQL to have multiple mutations side by side as CTEs.

Is it anywhere valid to *execute* two CTE mutations that do not depend on each other, concurrently?

The short answer is "it depends".

The long answer is, PostgreSQL does not do this. PostgreSQL executes its top level CTEs one after another, sequentially, so that multiple mutations never execute concurrently. So we can't just at pg and copy its concurrency rules. It doesn't have any.

If we make  up our own rules we'd need to be careful about two things:

- CTE semantics say "a mutation only happens once" (referential equivalence)
- if a 2nd mutation refers to the same table as the first, it should either see the data as if the other mutation had completely finished, or has not yet started (atomicity)

So I'd say we're free to execute mutations concurrently if we're able to prove during planning that they have no overlapping k/v ranges.

### What about EXPLAIN(PLAN)

The SQL syntax `EXPLAIN(PLAN) xxx` means "show the query plan for the SQL syntax `xxx`.

The result is a relational expression for a relation with 3 columns (tree, field, value) which contains a textual description of the query plan. **The query is not executed**.

Since that result is itself relational it's possible (and desirable) to have it in a sub-query, for example `select * from [explain ...]` or `with a as (explain ...) table a`.

This raises a new question: what happens if we apply EXPLAIN to a mutation, then also place that EXPLAIN in a sub-query? for example `select * from [explain update ...]`

The answer to this can be constructed from the following basic principle: **EXPLAIN defines its own, fresh, top level scope for query planning**.

From that principle alone we can derive the following rules:
- the syntax passed after EXPLAIN must be valid if it was pasted at the top level of a SQL prompt
- therefore, it does not permit correlated subqueries
- the EXPLAIN itself cannot be correlated with its surroundings
- during planning an EXPLAIN, the highest level at which _anything_ underneath can be hoisted, is the EXPLAIN itself
- the EXPLAIN itself is uncorrelated and thus can be hoisted at the top level of the surrounding query
- all the planning needed to produce the EXPLAIN results can be processed before the start of execution of the surrounding SQL query.

So the rules above about semantic constraints and hoisting continue to apply, unchanged:

- "no correlated mutation" -> nothing special
- "mutation hoisted to top level" -> "mutation hoisted to _current_ top level", which is going to be the nearest EXPLAIN if any
- ditto for hoisting other things that we also want to hoist: _uncorrelated_ subqueries are typically also hoisted to a top level, _and_ we want them to be "retained" under EXPLAIN, so they're going to be hoisted to EXPLAIN and not above that.
- if there are multiple nested EXPLAINs, for example `explain explain select 1`, we apply the planning recursively, using depth-first, post-order processing.

Then we have:

```sql
explain select * from [update ...]
```

becomes
```sql
explain with a as (update...) select * from a
```

and 
```sql
select * from [explain select * from [update ...]]
```

becomes
```
with a as (explain with b as (update ...) select * from b)
select * from a
```

### What about EXPLAIN(ANALYZE)

That's a tough(er) cookie. In contrast to EXPLAIN(PLAN), EXPLAIN(ANALYSE) needs to run its query parameter.

I don't have a clear opinion of this, but I think we'd get some mileage by sticking to the following rules:

- EXPLAIN(ANALYZE) must be plannable like EXPLAIN(PLAN), i.e. no uncorrelated query underneath
- guarantee semantically that the query it's given is only executed once and the results are stored.

This way we can always hoist EXPLAIN(ANALYZE), run the explains sequentially prior to the execution of the main query, and deliver the results of EXPLAIN via a spool node or temp table.

**Comment by knz** [comment]
**Note: at this time I do not have any evidence that CockroachDB does something different from what is suggested above** - this issue is explanatory (and perhaps a guide to define more unit tests), not an outline of a discovered problem/bug.

Random thought: I'm thinking about turning this into a blog post or doc page, so that users can better learn how to reason about our query planning.

**Comment by RaduBerinde** [comment]
I think as far as the docs go, we should discourage `[ ... ]` and mention `WITH` should be used in all cases. We should of course document the semantics of `[ ... ]` somewhere but I wouldn't make it part of a blog post.

IMO we should strive to allow what PG provides and not more (meaning we would also disallow non-top-level CTEs with mutations). The `[ <mutation> ]` case would be handled specially (for backward compatibility): the mutation is not allowed to reference anything outside of the brackets (i.e. outer columns or other CTEs), and we always pull it up into a top-level CTE.

In fact, I think we should define and document `[ ... ]` as syntactic sugar for a top-level WITH in all cases (and implement the semantics that follow from that).

@justinj is currently working on these things.

**Comment by RaduBerinde** [comment]
> EXPLAIN defines its own, fresh, top level scope for query planning.

Agreed, this is something Justin pointed out recently. CREATE VIEW is similar.

**Comment by andy-kimball** [comment]
Here is a related issue I filed, since we should be forbidding correlated mutations: https://github.com/cockroachdb/cockroach/issues/41074

**Comment by andy-kimball** [comment]
Also, as @RaduBerinde points out, `CREATE VIEW` has similar problems when mixed with the `[ ... ]` syntax. @knz, can you think of any other cases that would need to establish a new "top-level" context?

**Comment by knz** [comment]
Responding to @RaduBerinde above:

> the mutation is not allowed to reference anything outside of the brackets (i.e. outer columns or other CTEs), and we always pull it up into a top-level CTE.

The rules I spell out above are clear - if the `[...]` is a mutation it must be uncorrelated and semantically considered to be executed just once at the top level, and I don't think we've ever set expectations otherwise.

Responding to @andy-kimball 

> can you think of any other cases that would need to establish a new "top-level" context?

yes a couple come to mind:

- EXPORT FROM ...
- CREATE TABLE AS ...
- query clauses listed in PARTITION VALUES (...) of CREATE
- query clauses in CHECK and computed expressions in CREATE and ALTER TABLE
- ALTER TABLE/INDEX SPLIT AT ...
- CANCEL/SHOW/PAUSE/RESUME JOBS/QUERIES/SESSIONS ...
- PREPARE (although that one forms its own long winded story)

In general any relational node which _currently_ requires its own top-level execution processor needs to be restricted to apply to non-correlated relational clauses only.

(The discussion is a bit muddied, perhaps, by the observation that these nodes have an empty relational _result_ and thus are _currently_ not allowed as CTE. However, I think that observation is a red herring. The hoisting rules should consider them as blockers for hoisting without having to check whether they are at the top level or not.)

**Comment by knz** [comment]
To come back to this topic:

> the mutation is not allowed to reference anything outside of the brackets (i.e. outer columns or other CTEs), and we always pull it up into a top-level CTE.

I think we'd probably also be OK to require that _anything_ within `[...]` (mutation or not) be uncorrelated. This may allow us to turn `[...]` into `with` even before scope analysis.

**Comment by knz** [comment]
There's another special thing about CREATE VIEW: the AST that is given to it should be planned, but the plan discarded after it's been validated.

 That's because the AST should be stored as-is in the view descriptor and the execution of CREATE VIEW does not need the plan for its relational argument.

**Comment by RaduBerinde** [comment]
> There's another special thing about CREATE VIEW: the AST that is given to it should be planned, but the plan discarded after it's been validated.

Indeed, that's what [happens](https://github.com/cockroachdb/cockroach/blob/62f63c57a835a91c9119e8598c7b8a68099d2fb5/pkg/sql/opt/optbuilder/create_view.go#L42).

**Comment by RaduBerinde** [comment]
> I think we'd probably also be OK to require that anything within [...] (mutation or not) be uncorrelated. This may allow us to turn [...] into with even before scope analysis.

Yes, that's what I was suggesting when I said we should define `[...]` as syntactic sugar for a top-level WITH in all cases.

**Comment by andy-kimball** [comment]
Yes, I like that option. We've got no indication that anyone's using correlated `[...]` expressions and we've got more important things to support instead. Radu, can you or Justin add a fix to 19.2 to error if a `[...]` expression is correlated or contains mutations? If we wait another release to prohibit this, it increases chance that someone starts depending on it.

**Comment by knz** [comment]
Replying to @RaduBerinde and @andy-kimball's last two comments.

It seems to me that you're attempting to create a focus on special properties of the `[...]` syntax, by suggesting to disallow something about it (correlation) by a special-purpose check.

I'd like to more strongly remind you that this is neither desirable nor necessary. As I have written at the top, the semantic analysis would be best served by desugaring any use of `[...]` by a CTE definition, _prior to scope analysis_, and then proceeding from there in optbuilder as if the CTE was there from the beginning.

We can then discuss the choice of desugaring:

1. we can desugar at the same syntax level, e.g. turn `select * from [xxx]` into `select * from (with a as (xxx) table a)`
2. or, we can desugar at the top level, e.g. turn `select * from [xxx]` into `with a as (xxx) select * from a`

The discussion above has revealed to us that both approaches are valid.

Now as to which one is desirable. I have a strong hunch that approach 2 is the best, for a couple of reasons:

1. it is simpler to implement
2. it avoids the extra work to hoist the CTE to the top level recursively afterwards
3. if a user mistakenly typed in correlated SQL in there it will be found invalid after the desugaring into a top-level CTE, giving us the restriction you want "for free" (no separate/new check necessary).

**Comment by RaduBerinde** [comment]
@knz yes @justinj is trying out approach 2 for 20.1. @andy-kimball was saying we should add a special check to 19.2 so that no new correlated uses pop up until 20.1.

**Comment by knz** [comment]
thanks for explaining, I had misunderstood

**Comment by RaduBerinde** [comment]
Approach 2 has been implemented.
